### PR TITLE
Added support for FHIR R4 4.0.1 

### PR DESCRIPTION
**Pull request by AndrewBird81**
Also updated version from 1.3.1 -> 1.3.2 and bumped lodash version.

I tried to mimic the existing code, so there is duplication in the unit test. The most noticeable difference between 4.0.0 and 4.0.1 is the structure of the `Extension` type. In the 4.0.0 modelinfo.xml, the `url` element is defined as:
'<ns4:element` name="url" elementType="System.String"/>`

But in 4.0.1, we have:
`<ns4:element name="url" elementType="FHIR.uri"/>`

As a consequence, an `Extension`'s `url` is no longer a simple string, but a complex object with its own `value`, `id`, `url`, and `Extension` array.

### CODE REVIEW

**Comment by brynrhodes** [comment]
So, this is interesting, we had to make a specific exception for the URL attribute in the ModelInfo generation logic because the url element in HAPI is actually a [URIType](https://github.com/hapifhir/org.hl7.fhir.core/blob/master/org.hl7.fhir.r4/src/main/java/org/hl7/fhir/r4/model/Extension.java#L102)

We could address this in the model info and update the Java provider to account for the discrepancy in the Data Access layer. Thoughts?

**Comment by AndrewBird81** [comment]
Apologies @brynrhodes , but I'm not understanding where the issue lies. Are you referring to the mismatch between the name `url` and the type `uri`? Or are you saying the `url` field in FHIR Extensions should be a System.String, as it was in the 4.0.0 modelinfo.xml? Or is it something else entirely that I'm missing?

File: test/r401_test.js
Comment (hossenlopp) [inline test/r401_test.js]: Update this describe be different than the 4.0.0 tests. In the current state it's a bit confusing to know which test results are for which FHIR version.

**Review by hossenlopp — CHANGES_REQUESTED** [review]
👍 ModelInfo file matches the version at https://github.com/cqframework/clinical_quality_language/blob/master/Src/java/quick/src/main/resources/org/hl7/fhir/fhir-modelinfo-4.0.1.xml

EXM104 Bundle file fixture is not used by any tests. Perhaps remove it.

Why are there changes in the DSTU2, STU3 and R4 test files?

**Comment by brynrhodes** [comment]
@AndrewBird81 , sorry for the confusion, yes, the 4.0.1 model info is correct based on the published spec, I thought we had this workaround, but the workaround was actually for 4.0.0 and it was corrected in 4.0.1. Again, sorry for the confusion, disregard. :)

File: test/r401_test.js
Comment (AndrewBird81) [reply test/r401_test.js]: Good point- I updated both descriptions to avoid confusion.

Comment (AndrewBird81) [reply test/r401_test.js]: I also removed the unused test fixture. As for the changes to the other unit tests, they were importing the same files multiple times, so I refactored them to be more efficient. I've grown accustomed to the ES6 and TypeScript modules, so seeing dynamic requires has begun to look like an anti-pattern to me. But I could revert them if you prefer.

Comment (hossenlopp) [reply test/r401_test.js]: I think that makes sense. But a possible downside is that if a test ends up being destructive to the data loaded it could affect other tests. I think that could have been why it was done in the `beforeEach`. I'll leave this to @cmoesel if he thinks this refactoring is ok.

File: package.json
Comment (cmoesel) [inline package.json]: Since this adds support for a new model (FHIR 4.0.1), I'd consider that a _minor_ release rather than a _patch_ release; it now supports entire CQL libraries it did not support before.  So I'd suggest we bump to 1.4.0.  Would you agree?

File: test/r401_test.js
Comment (cmoesel) [inline test/r401_test.js]: These patients are actually Luna and Johnnie, so we should rename the variables to `patientLuna` and `patientJohnnie`.

File: test/r4_test.js
Comment (cmoesel) [inline test/r4_test.js]: These patients are actually Luna and Johnnie, so we should rename the variables to `patientLuna` and `patientJohnnie`.

File: test/r401_test.js
Comment (cmoesel) [inline test/r401_test.js]: Can you update this to:
```js
describe('#R4 v4.0.1', () => {
```
like you did for the FHIRWrapper test above?  That way it's easier to distinguish failing 4.0.1 tests from failing 4.0.0 tests.

File: test/r401_test.js
Comment (cmoesel) [inline test/r401_test.js]: I don't think the 4.0.1 modelinfo has this legacy, so it's not important for 4.0.1 to support this.  We can probably remove this test.

File: test/r401_test.js
Comment (AndrewBird81) [reply test/r401_test.js]: For what it's worth, they're already modified in the `before()` in a safe way:
`conditionResourceWithNoType = JSON.parse(JSON.stringify(conditionResource));`
`delete conditionResourceWithNoType.resourceType;`

File: test/r401_test.js
Comment (cmoesel) [inline test/r401_test.js]: Oops.  This should probably say "R4" instead of "STU3".  I'm guessing this same typo existing in the 4.0.0 test suite as well (and that you merely copy/pasted it).

File: test/r401_test.js
Comment (cmoesel) [inline test/r401_test.js]: Similarly, the 4.0.1 modelinfo has `primaryCodePath` `medication`, so I don't think this test is relevant for 4.0.1.

File: test/r4_test.js
Comment (cmoesel) [inline test/r4_test.js:9]: Adding `v4.0.0` to the suite name is good.  Can you do the same for the other `describe` in this file?

**Review by cmoesel — CHANGES_REQUESTED** [review]
Thanks for this, @AndrewBird81!  It will be great to have support for FHIR 4.0.1!  I've requested a few small changes -- nothing major at all.

As for the slight refactor in the existing tests, I'm fine with eliminating the repeated calls to `require` in the `beforeEach` clauses.  I suspect that maybe at one point we were modifying the definitions?  I don't know -- but I'm fine w/ the change.  And if the _production_ code mutated the patient data, I'd want to know about that!  So I guess that in a weird way this ends up testing for that too?

**Comment by AndrewBird81** [comment]
@cmoesel - excellent feedback! I believe I've made the requested changes. Kindly let me know if you find anything else. Thanks again!

**Comment by cmoesel** [comment]
Thanks, @AndrewBird81.  I'd like to manually test this w/ some CQL and R4 patients later today before I officially approve.

**Review by cmoesel — APPROVED** [review]
This looks great!  I've run it manually with some FHIR 4.0.1 CQL and also confirmed it there.  Thanks for the contribution, @AndrewBird81!

**Review by hossenlopp — APPROVED** [review]
Thanks for doing this work, @AndrewBird81!

**Comment by cmoesel** [comment]
This has now been released to NPM as version 1.4.0.

GitHub: https://github.com/cqframework/cql-exec-fhir/releases/tag/v1.4.0
NPM: https://www.npmjs.com/package/cql-exec-fhir/v/1.4.0

**Pull request merged**
